src/twisted/application/runner/_pidfile.py

```diff
@@ -0,0 +1,270 @@
+        @raise EnvironmentError: If this PID file cannot be read.
+        @raise ValueError: If this PID file's content is invalid.
+        """
+
+
+    def write(pid):
+        """
+        Store a PID in this PID file.
+
+        @param pid: A PID to store.
+        @type pid: L{int}
+
+        @raise EnvironmentError: If this PID file cannot be written.
+        """
+
+
+    def writeRunningPID():
+        """
+        Store the PID of the current process in this PID file.
+
+        @raise EnvironmentError: If this PID file cannot be written.
+        """
+
+
+    def remove():
+        """
+        Remove this PID file.
+
+        @raise EnvironmentError: If this PID file cannot be removed.
+        """
+
+
+    def isRunning():
+        """
+        Determine whether there is a running process corresponding to the PID
+        in this PID file.
+
+        @return: True if this PID file contains a PID and a process with that
+        PID is currently running; false otherwise.
+        @rtype: L{bool}
+
+        @raise EnvironmentError: If this PID file cannot be read.
+        @raise InvalidPIDFileError: If this PID file's content is invalid.
+        @raise StalePIDFileError: If this PID file's content refers to a PID
+            for which there is no corresponding running process.
+        """
+
+
+    def __enter__():
+        """
+        Enter a context using this PIDFile.
+
+        Writes the PID file with the PID of the running process.
+
+        @raise AlreadyRunningError: A process corresponding to the PID in this
+            PID file is already running.
+        """
+
+
+    def __exit__(exc_type, exc_value, traceback):
+        """
+        Exit a context using this PIDFile.
+
+        Removes the PID file.
+        """
+
+
+
+@implementer(IPIDFile)
+class PIDFile(object):
+    """
+    Concrete implementation of L{IPIDFile} based on C{IFilePath}.
+    """
+
+    _log = Logger()
+
+
+    @staticmethod
+    def format(pid):
+        """
+        Format a PID file's content.
+
+        @param pid: A process ID.
+        @type pid: int
+
+        @return: Formatted PID file contents.
+        @rtype: L{bytes}
+        """
+        return u"{}\n".format(int(pid)).encode("utf-8")
+
+
+    def __init__(self, filePath):
+        """
+        @param filePath: The path to the PID file on disk.
+        @type filePath: L{IFilePath}
+        """
+        self.filePath = filePath
+
+
+    def read(self):
+        pidString = b""
+        try:
+            with self.filePath.open() as fh:
+                for pidString in fh:
+                    break
+        except OSError as e:
+            if e.errno == errno.ENOENT:  # No such file
+                raise NoPIDFound("PID file does not exist")
+            raise
+
+        try:
+            return int(pidString)
+        except ValueError:
+            raise InvalidPIDFileError(
+                "non-integer PID value in PID file: {!r}".format(pidString)
```

Comment:
Including the path name of the PID file in question would be super useful for diagnostics.